stdlib/public/core/StringUTF16View.swift

```diff
@@ -517,7 +518,180 @@ extension _StringGuts {
+    
+    while readPtr + MemoryLayout<U>.stride < endPtr {
+      //Find the number of continuations (0b10xxxxxx)
+      let sValue = readPtr.load(as: S.self)
+      let continuations = S.zero.replacing(with: S.one, where: sValue .< -65 + 1)
+      let continuationCount = Int(continuations.wrappedSum())
+            
+      //Find the number of 4 byte code points (0b11110xxx)
+      let uValue = readPtr.load(as: U.self)
+      let fourBytes = U.zero.replacing(with: U.one, where: uValue .>= 0b11110000)
+      let fourByteCount = Int(fourBytes.wrappedSum())
+            
+      utf16Count &+= (U.scalarCount - continuationCount) + fourByteCount
+      
+      readPtr += MemoryLayout<U>.stride
+    }
+    
+    return utf16Count
+  }
+  
+  @inline(__always)
+  internal func _utf16Distance(from start: Index, to end: Index) -> Int {
+    _internalInvariant(end.transcodedOffset == 0 || end.transcodedOffset == 1)
+    return (end.transcodedOffset - start.transcodedOffset) + _guts.withFastUTF8(
+      range: startIndex._encodedOffset ..< end._encodedOffset
+    ) { utf8 in
+      let rawBuffer = UnsafeRawBufferPointer(utf8)
+      guard rawBuffer.count > 0 else { return 0 }
+      
+      var utf16Count = 0
+      var readPtr = rawBuffer.baseAddress.unsafelyUnwrapped
+      let initialReadPtr = readPtr
+      let endPtr = readPtr + rawBuffer.count
+      
+      //eat leading continuations
+      while readPtr < endPtr {
+        let byte = readPtr.load(as: UInt8.self)
+        if !UTF8.isContinuation(byte) {
+          break
+        }
+        readPtr += 1
+      }
+      
+     // align
+      while readPtr < endPtr
+        && UInt(bitPattern: readPtr) % UInt(MemoryLayout<SIMD8<UInt8>>.stride) != 0 {
+        let byte = readPtr.load(as: UInt8.self)
+        let len = _utf8ScalarLength(byte)
+        // if we don't have enough bytes left, we don't have a complete scalar,
+        // so don't add it to the count.
+        if readPtr + len <= endPtr {
+          utf16Count &+= len == 4 ? 2 : 1
+        }
+        readPtr += len
+      }
+      
+    //  utf16Count &+= _utf16Length(
+    //    readPtr: &readPtr,
+    //    endPtr: endPtr,
+    //    unsignedSIMDType: SIMD64<UInt8>.self,
+    //    signedSIMDType: SIMD64<Int8>.self
+    //  )
+    //
+    //  utf16Count &+= _utf16Length(
+    //    readPtr: &readPtr,
+    //    endPtr: endPtr,
+    //    unsignedSIMDType: SIMD32<UInt8>.self,
+    //    signedSIMDType: SIMD32<Int8>.self
+    //  )
+    //
+    //  utf16Count &+= _utf16Length(
+    //    readPtr: &readPtr,
+    //    endPtr: endPtr,
+    //    unsignedSIMDType: SIMD16<UInt8>.self,
+    //    signedSIMDType: SIMD16<Int8>.self
+    //  )
+
+      // Currently, using SIMD sizes above SIMD8 is slower
+      utf16Count &+= _utf16Length(
+        readPtr: &readPtr,
+        endPtr: endPtr,
+        unsignedSIMDType: SIMD8<UInt8>.self,
+        signedSIMDType: SIMD8<Int8>.self
+      )
+    //
+    //  utf16Count &+= _utf16Length(
```

Comment:
Do we want smaller vectors for tail cleanup or just go straight to byte by byte?